src/data/complex/exponential.lean

```diff
@@ -420,48 +422,165 @@ begin
+eq_conj_iff_re.1 $ by rw [← sinh_conj, conj_of_real]
+
+@[simp] lemma of_real_sinh (x : ℝ) : (real.sinh x : ℂ) = sinh x :=
+of_real_sinh_of_real_re _
+
+@[simp] lemma sinh_of_real_im (x : ℝ) : (sinh x).im = 0 :=
+by rw [← of_real_sinh_of_real_re, of_real_im]
+
+lemma sinh_of_real_re (x : ℝ) : (sinh x).re = real.sinh x := rfl
+
+lemma cosh_conj : cosh (conj x) = conj (cosh x) :=
+by rw [cosh, ← conj_neg, exp_conj, exp_conj, ← conj_add, cosh, conj_div, conj_two]
+
+@[simp] lemma of_real_cosh_of_real_re (x : ℝ) : ((cosh x).re : ℂ) = cosh x :=
+eq_conj_iff_re.1 $ by rw [← cosh_conj, conj_of_real]
+
+@[simp] lemma of_real_cosh (x : ℝ) : (real.cosh x : ℂ) = cosh x :=
+of_real_cosh_of_real_re _
+
+@[simp] lemma cosh_of_real_im (x : ℝ) : (cosh x).im = 0 :=
+by rw [← of_real_cosh_of_real_re, of_real_im]
+
+lemma cosh_of_real_re (x : ℝ) : (cosh x).re = real.cosh x := rfl
+
+lemma tanh_eq_sinh_div_cosh : tanh x = sinh x / cosh x := rfl
+
+@[simp] lemma tanh_zero : tanh 0 = 0 := by simp [tanh]
+
+@[simp] lemma tanh_neg : tanh (-x) = -tanh x := by simp [tanh, neg_div]
+
+lemma tanh_conj : tanh (conj x) = conj (tanh x) :=
+by rw [tanh, sinh_conj, cosh_conj, ← conj_div, tanh]
+
+@[simp] lemma of_real_tanh_of_real_re (x : ℝ) : ((tanh x).re : ℂ) = tanh x :=
+eq_conj_iff_re.1 $ by rw [← tanh_conj, conj_of_real]
+
+@[simp] lemma of_real_tanh (x : ℝ) : (real.tanh x : ℂ) = tanh x :=
+of_real_tanh_of_real_re _
+
+@[simp] lemma tanh_of_real_im (x : ℝ) : (tanh x).im = 0 :=
+by rw [← of_real_tanh_of_real_re, of_real_im]
+
+lemma tanh_of_real_re (x : ℝ) : (tanh x).re = real.tanh x := rfl
+
+lemma cosh_add_sinh : cosh x + sinh x = exp x :=
+by rw [← domain.mul_left_inj (@two_ne_zero' ℂ _ _ _), mul_add,
+       two_cosh, two_sinh, add_add_sub_cancel, two_mul]
+
+lemma sinh_add_cosh : sinh x + cosh x = exp x :=
+by rw [add_comm, cosh_add_sinh]
+
+lemma cosh_sub_sinh : cosh x - sinh x = exp (-x) :=
+by rw [← domain.mul_left_inj (@two_ne_zero' ℂ _ _ _), mul_sub,
+       two_cosh, two_sinh, add_sub_sub_cancel, two_mul]
+
+lemma cosh_sq_sub_sinh_sq : cosh x ^ 2 - sinh x ^ 2 = 1 :=
+by rw [sq_sub_sq, cosh_add_sinh, cosh_sub_sinh, ← exp_add, add_neg_self, exp_zero]
 
 @[simp] lemma sin_zero : sin 0 = 0 := by simp [sin]
 
 @[simp] lemma sin_neg : sin (-x) = -sin x :=
 by simp [sin, exp_neg, (neg_div _ _).symm, add_mul]
 
+lemma two_sin : 2 * sin x = (exp (-x * I) - exp (x * I)) * I :=
+mul_div_cancel' _ two_ne_zero'
+
+lemma two_cos : 2 * cos x = exp (x * I) + exp (-x * I) :=
+mul_div_cancel' _ two_ne_zero'
+
+lemma sinh_mul_I : sinh (x * I) = sin x * I :=
+by rw [← domain.mul_left_inj (@two_ne_zero' ℂ _ _ _), two_sinh,
+       ← mul_assoc, two_sin, mul_assoc, I_mul_I, mul_neg_one,
+       neg_sub, neg_mul_eq_neg_mul]
+
+lemma cosh_mul_I : cosh (x * I) = cos x :=
+by rw [← domain.mul_left_inj (@two_ne_zero' ℂ _ _ _), two_cosh,
+       two_cos, neg_mul_eq_neg_mul]
+
 lemma sin_add : sin (x + y) = sin x * cos y + cos x * sin y :=
-begin

```

Comment:
This could actually be made even shorter by using `sinh_add`

Reply:
But I did use `sinh_add`?

Reply:
Sorry. I didn't notice